rfcs/20210510-binary-size-reduction/README.md

```diff
@@ -0,0 +1,156 @@
+# oneDNN binary size reduction through build system.
+
+## Problem Statement.
+Several oneDNN users due to their usage model specifics would love to see their
+application linked statically with oneDNN (or shared object of the library
+itself) doesn't exceed specific disk space quota. One of such oneDNN users is
+OpenVINO toolkit. Our experiments showed that binary size of statically linked
+end-user application may take up to 64 MB [1]. Further analysis revealed several
+spots that contribute most:
+
+* The implementation of threading layer through templates. This is TBB specific,
+  OMP doesn't suffer from it. Takes ~14 MB.
+* The implementation of reorders through templates of plain C code and their
+  multiple instantiations. Takes ~9 MB.
+* The implementation of GEMM autogenerated JIT-ted kernels. Takes ~14 MB.
+* The implementation of convolution kernels apart of GEMM. Takes ~8 MB.
+* Rest functionality takes ~12 MB + ~2 MB of core API.
+
+## Background
+The reason why static linking can't solve the problem itself is the way library
+API is implemented. The call to `dnnl_primitive_desc_create` (which is also used
+underneath correspondent C++ API) relies on statically defined implementation
+lists which contain all primitives at ones. That means that if user calls the
+function (which is an essential part of the API), they get the full library in
+their application. And there are not many choices how to deal with such problem.
+One is to switch off implementations in compile time (this RFC proposal) or
+using a different API which would separate things on different levels (at least
+on primitive one).
+
+## Proposal
+To control the size, the proposal is to utilize the power of build system and
+let the user specify set of desired options. The idea is to have two sets of
+macros, one is external and would be defined by the build system based on user
+choices. This external set would control another set - internal - which lives
+inside the library and modifies implementation lists. Each primitive would get
+several macros to switch on/off certain implementations in compile time. Further
+optimizations of disk space will be provided by linker which shouldn't include
+unused code inside the end-user application unless certain scenario happens [2].
+
+As a part of this proposal, there are only two possible options to specify right
+now:
+* Propagation kind: inference or training (the default). Usage example, which
+  enables inference only:
+  - -DNNL_ENABLE_PROP_KIND=INFERENCE
+* Primitives. Usage example, which enables only convolution and matmul:
+  - -DDNNL_ENABLE_PRIMITIVES=CONVOLUTION;MATMUL
+
+This list may be extended in future upon request with data types and, possibly,
+ISA, if it's proven to help with further size reduction with minimal amount of
+enabling efforts and maintenance cost.
+
+Pros:
+1. No API changes and transparent enabling for the end-user.
```

Comment:
I'm not sure this can be implemented smoothly without leaking of new macros to other layers which I would suggest to avoid. Another thing is it may not be that useful since user configures build on their own and should know what was disabled and what's not.

Reply:
@igorsafo, what benefits do you think the new statuses will give the users? I mean, suppose primitive descriptor returns `dnnl_disabled` instead of `dnnl_unimplemented`, do we expect users to have different error handling for those two?

Reply:
Yes, unimplemented might be caused by unsupported post op in attributes or unsupported architecture for the given data type configuration etc, while `dnnl_disabled' means that this functionality is completely disabled. `dnnl_unimplemented` could be workarounded , for example, by removing post ops from attributes. It might be implemented in another way, but it would be helpful for user to have an API to understand whether functionality is enabled or not. For example, for GPU we have DNNL_GPU_RUNTIME.

Reply:
The mechanism you describe doesn't seem to need knowledge about that. I think about the described workflow as about  a "try approach" that is, there are a number of attempts to create a primitive descriptor (w/ or w/o post ops, etc) and if all of them failed the users will have to provide a fallback. The only benefit seems to be the ability to use fallback after the first attempt failed, but I'm not sure how useful that would be.

Reply:
Sounds like resolving to me. Feel free to open if there's something to re-iterate here.